### PR TITLE
linked time: stepIndex moving: add function to collect all step values within a image card

### PR DESCRIPTION
**Pull request by japie1235813**
Creates a function to get step values from a image card. This prepares for stepIndex moving. Motivation and full context please see #5615.

### CODE REVIEW

File: tensorboard/webapp/metrics/store/metrics_store_internal_utils_test.ts [hunk 1]
```diff
@@ -639,4 +645,84 @@ describe('metrics store utils', () => {
...
+      images: {runToLoadState: {}, runToSeries: {}},
+    };
+    let timeSeriesData: TimeSeriesData = {
+      scalars: {},
+      histograms: {},
+      images: {},
+    };
+
+    beforeEach(() => {
+      loadables = {
```
Comment (bmd3k) [inline tensorboard/webapp/metrics/store/metrics_store_internal_utils_test.ts:671]: I don't think there is any reason for `loadables` to exist? You can just create inline in timeSeriesData? The extra level of indirection is very confusing for understanding the tests.

File: tensorboard/webapp/metrics/store/metrics_store_internal_utils.ts [hunk 2]
```diff
@@ -374,11 +379,41 @@ export function canCreateNewPins(state: MetricsState) {
...
+  previousCardStepIndex: CardStepIndexMap,
+  cardMetadataMap: CardMetadataMap,
+  timeSeriesData: TimeSeriesData
 ): CardStepIndexMap {
   const nextCardStepIndex = {...previousCardStepIndex};
 
+  Object.keys(previousCardStepIndex).forEach((cardId) => {
+    if (!cardId.includes('"plugin":"images"')) return;
+
+    const stepValues = getImageCardStepValues(
```
[the remark this answers is not in the document]
Comment (japie1235813) [reply tensorboard/webapp/metrics/store/metrics_store_internal_utils.ts]: Removed.

File: tensorboard/webapp/metrics/store/metrics_store_internal_utils.ts [hunk 3]
```diff
@@ -374,11 +379,41 @@ export function canCreateNewPins(state: MetricsState) {
...
+  const loadable = getTimeSeriesLoadable(timeSeriesData, plugin, tag, sample);
+  const runToSeries = loadable ? loadable.runToSeries : null;
+  const timeSeries =
+    runToSeries && runId && runToSeries.hasOwnProperty(runId)
+      ? (runToSeries[runId] as ImageStepDatum[])
+      : [];
+  return timeSeries.map((stepDatum) => stepDatum.step);
+}
+
+export const TEST_ONLY = {getImageCardStepValues, util};
```
[the remark this answers is not in the document]
Comment (japie1235813) [reply tensorboard/webapp/metrics/store/metrics_store_internal_utils.ts:407]: It's from previous code. Checked that it's used in `canCreateNewPins` function

File: tensorboard/webapp/metrics/store/metrics_store_internal_utils.ts [hunk 4]
```diff
@@ -374,11 +379,41 @@ export function canCreateNewPins(state: MetricsState) {
...
+ * Returns step values of a image card.
+ */
+export function getImageCardStepValues(
+  cardId: string,
+  cardMetadataMap: CardMetadataMap,
+  timeSeriesData: TimeSeriesData
+): number[] {
+  const {plugin, tag, sample, runId} = cardMetadataMap[cardId];
+  const loadable = getTimeSeriesLoadable(timeSeriesData, plugin, tag, sample);
+  const runToSeries = loadable ? loadable.runToSeries : null;
```
[the remark this answers is not in the document]
Comment (japie1235813) [reply tensorboard/webapp/metrics/store/metrics_store_internal_utils.ts]: Thanks! Modified as suggested.

File: tensorboard/webapp/metrics/store/metrics_store_internal_utils_test.ts [hunk 5]
```diff
@@ -639,4 +645,84 @@ describe('metrics store utils', () => {
...
+        histograms: {tagB: loadables.histograms},
+        images: {tagC: {0: loadables.images}},
+      };
+    });
+
+    it(`gets empty step value when no run id in time series data`, () => {
+      expect(
+        getImageCardStepValues(cardId, cardMetadataMap, timeSeriesData)
+      ).toEqual([]);
+    });
```
[the remark this answers is not in the document]
Comment (japie1235813) [reply tensorboard/webapp/metrics/store/metrics_store_internal_utils_test.ts:689]: based on the type definition
```
export type CardMetadataMap = Record<
  NonPinnedCardId | PinnedCardId,
  CardMetadata
>;

export interface CardMetadata {
  ...
  /**
   * A `null` runId indicates all runs.
   */
  runId: string | null;
}
```
In annotation it states If run id is  null, it means all runs. Practically it does not make sense to me that runId is null for an image card.

If this helps with explanation, this part of code is referencing image_container here
https://cs.opensource.google/tensorflow/tensorboard/+/master:tensorboard/webapp/metrics/views/card_renderer/image_card_container.ts;l=190-196?q=image_card%20&ss=tensorflow%2Ftensorboard

[the remark this answers is not in the document]
Comment (japie1235813) [reply tensorboard/webapp/metrics/store/metrics_store_internal_utils_test.ts:689]: For 2., I was trying to mock `getTimeSeriesLoadable` but failed (it's global function, similar to what we had discussed but didn't resolve case before)

I do think it's possible that it returns null. For now I can only crippled cardMetadataMap and thus timeSeriesData to make it return null.

**Review by bmd3k — APPROVED** [review]

**Pull request merged**